src/DependencyInjection/Compiler/RegisterFragmentsPass.php

```diff
@@ -0,0 +1,171 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * Copyright (c) 2005-2017 Leo Feyer
+ *
+ * @license LGPL-3.0+
+ */
+
+namespace Contao\CoreBundle\DependencyInjection\Compiler;
+
+use Contao\CoreBundle\Fragment\FragmentConfig;
+use Contao\CoreBundle\Fragment\FragmentPreHandlerInterface;
+use Contao\CoreBundle\Fragment\Reference\ContentElementReference;
+use Contao\CoreBundle\Fragment\Reference\FrontendModuleReference;
+use Symfony\Component\Config\Definition\Exception\InvalidConfigurationException;
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\Compiler\PriorityTaggedServiceTrait;
+use Symfony\Component\DependencyInjection\Container;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Definition;
+use Symfony\Component\DependencyInjection\Reference;
+
+/**
+ * Registers Contao fragments in the registry.
+ * For custom fragment tags, create your own compiler pass by extending this class and replacing the process() method.
+ */
+class RegisterFragmentsPass implements CompilerPassInterface
+{
+    use PriorityTaggedServiceTrait;
+
+    /**
+     * Adds the fragments to the registry.
+     *
+     * @param ContainerBuilder $container
+     */
+    public function process(ContainerBuilder $container): void
+    {
+        if (!$container->has('contao.fragment.registry')) {
+            return;
+        }
+
+        $this->registerFragments($container, ContentElementReference::TAG_NAME);
+        $this->registerFragments($container, FrontendModuleReference::TAG_NAME);
+    }
+
+    /**
+     * Registers the fragments.
+     *
+     * @param ContainerBuilder $container
+     * @param string           $tag
+     *
+     * @throws InvalidConfigurationException
+     */
+    protected function registerFragments(ContainerBuilder $container, string $tag): void
```

Comment:
Do all of the methods have to be protected? We should make those methods private, which are unlikely to be required to be overwritable.

Reply:
Yes they are intentionally protected because other people need to extend the compiler pass to register their own fragments. See the class description.

Reply:
I have read the class description – see my comment. 😄 

> We should make those methods private, which are unlikely to be required to be overwritable

Are all methods likely to be overwritten?

Reply:
Ah, yeah. They are all likely to be used/overwritten in my opinion. 

- `registerFragments` is the main entry point
- `getFragmentConfig` is necessary if you have custom options in your tag configuration
- `getControllerName` if your controllers work differently
- `addPreHandlers` to alter prehandler registration
- `getFragmentType` to alter suffix or similar stuff when generating the type